File: Airport.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Lab4
{

    public class Airport
    {
        public string Code { get; set; }
        public string Name { get; set; }
        public string City { get; set; }
        public string Country { get; set; }

        public Airport() { }

        public Airport(string code, string name, string city, string country)
        {
            Code = code;
            Name = name;
            City = city;
            Country = country;
        }

        public static List<Airport> LoadAll()
        {
            var airports = new List<Airport>();
            if (!File.Exists("data/airports.txt"))
                return airports;

            foreach (var line in File.ReadAllLines("data/airports.txt"))
            {
                if (string.IsNullOrWhiteSpace(line) || line.StartsWith("#")) continue;

                var parts = Utils.Split(line, ';');
                if (parts.Count != 4) continue;

                airports.Add(new Airport(parts[0], parts[1], parts[2], parts[3]));
            }

            return airports;
        }

        public static void SaveAll(List<Airport> airports)
        {
            Utils.EnsureDataDir();
            using var writer = new StreamWriter("data/airports.txt", false);
            writer.WriteLine("# Code;Name;City;Country");
            foreach (var a in airports)
            {
                writer.WriteLine($"{a.Code};{a.Name};{a.City};{a.Country}");
            }
        }
    }
}

File: Airport.cpp
#include "Airport.h"
#include "Utils.h"
#include <fstream>
#include <iostream>
#include <algorithm>

int Airport::totalAirports = 0;

void Airport::validate(const std::string& code, const std::string& name) {
    if (code.empty() || code.size() > 5) {
        throw std::invalid_argument("Код аэропорта должен быть 1–5 букв.");
    }
    if (name.empty()) {
        throw std::invalid_argument("Название аэропорта не может быть пустым.");
    }
    if (!std::all_of(code.begin(), code.end(), [](unsigned char c) { return std::isalpha(c); })) {
        throw std::invalid_argument("Код аэропорта должен содержать только буквы (например, SVO).");
    }
}

Airport::Airport(const std::string& cd, const std::string& nm, const std::string& ct, const std::string& cn)
    : code(cd), name(nm), city(ct), country(cn) {
    validate(code, name);
    ++totalAirports;
}

bool Airport::operator==(const Airport& other) const {
    return this->code == other.code;
}

bool Airport::operator!=(const Airport& other) const {
    return !(*this == other);
}

std::ostream& operator<<(std::ostream& os, const Airport& a) {
    os << "Аэропорт[" << a.code << " – " << a.name << ", " << a.city << ", " << a.country << "]";
    return os;
}

void Airport::loadAll(std::vector<std::unique_ptr<Airport>>& airports) {
    airports.clear();
    std::ifstream file("data/airports.txt");
    if (!file.is_open()) return;

    std::string line;
    while (std::getline(file, line)) {
        line.pop_back();
        if (line.empty() || line[0] == '#') continue;
        auto parts = split(line, ';');
        if (parts.size() != 4) continue;

        try {
            auto a = std::make_unique<Airport>(parts[0], parts[1], parts[2], parts[3]);
            airports.push_back(std::move(a));
        }
        catch (const std::exception& ex) {
            std::cerr << "Пропущен некорректный аэропорт:" << ex.what() << "\n";
        }
    }
    file.close();
}

void Airport::saveAll(const std::vector<std::unique_ptr<Airport>>& airports) {
    ensureDataDir();
    std::ofstream file("data/airports.txt");
    file << "# Code;Name;City;Country\n";
    for (const auto& a : airports) {
        file << a->code << ";"
            << a->name << ";"
            << a->city << ";"
            << a->country << "\n";
    }
    file.close();
}
